### PR TITLE
Patch upstream changes for modular block handler

### PR DESCRIPTION
**Pull request by rahulksnv**
Upstream PR: https://github.com/paritytech/substrate/pull/14014

Note: this would break compilation on subspace repo. https://github.com/subspace/subspace/pull/1393/ would need to be submitted to fix the build

### CODE REVIEW

**Review by nazar-pc — APPROVED** [review]
Makes sense

File: client/network/sync/src/block_request_handler.rs [hunk 1]
```diff
@@ -172,11 +181,15 @@ where
...
+		BlockRelayParams {
+			server: Box::new(Self { client, request_receiver, seen_requests }),
+			downloader: Arc::new(FullBlockDownloader::new(protocol_config.name.clone(), network)),
+			request_response_config: protocol_config,
+		}
 	}
 
 	/// Run [`BlockRequestHandler`].
-	pub async fn run(mut self) {
+	async fn process_requests(&mut self) {
```
Comment (nazar-pc) [inline client/network/sync/src/block_request_handler.rs:192]: Why did you change this from `&mut self` to `mut self`?

Comment (rahulksnv) [reply client/network/sync/src/block_request_handler.rs:192]: You meant the other way I guess. This is for size determination with the trait, otherwise:
```

error[E0161]: cannot move a value of type `dyn BlockServer<TBl>`
   --> client/service/src/builder.rs:793:3
    |
793 |         block_server.run().await;
    |         ^^^^^^^^^^^^^^^^^^ the size of `dyn BlockServer<TBl>` cannot be statically determined

```

Do you see an issue with this?

Comment (nazar-pc) [reply client/network/sync/src/block_request_handler.rs:192]: I was just curious in Subspace PR why it was necessary and it lead me here. No significant concerns in this case, but changing `mut self` to `&mut self` in general does have implications because often async functions are actually not always ready for this. If you had some accumulator before `while` then there would be a problem.

Comment (rahulksnv) [reply client/network/sync/src/block_request_handler.rs:192]: I see, can you give more details or example? Are you referring to capturing the reference in a closure?

Comment (nazar-pc) [reply client/network/sync/src/block_request_handler.rs:192]: It is generally referred to as cancellation safety: https://smallcultfollowing.com/babysteps/blog/2022/06/13/async-cancellation-a-case-study-of-pub-sub-in-mini-redis/
